### PR TITLE
Print all phrase combinations

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:**

## Feature Details
Adds the `--phrase <phrase>` command line argument which will print every possible result of the named phrase.
Also, did a little more refactoring and added some comments.

I would have preferred to handle all this in PrintData.cpp, but I need access to private members of Phrase, and I think making PrintData a friend of Phrase would be worse than just adding this new method to the Phrase class.

## Usage Examples
`EndlessSky.exe --phrase "friendly civilian"`

## Testing Done
`EndlessSky.exe --phrase "friendly civilian"` (I do not recommend this unless you have a chonky computer.)
`EndlessSky.exe --phrase "friendly militia 4"` (This is a lot safer.)

The militia result: 
[friendly militia 4.txt](https://github.com/endless-sky/endless-sky/files/10345551/friendly.militia.4.txt)

The file produced by asking for the friendly civilian hail is about 6 GB and I cannot even open it in VSCode.
So, for phrases with many sentences, it may be best to only do one at a time.

## Performance Impact
Running the phrase command does a lot of stuff, for large/complex phrases, the memory usage may be quite large, but it's fairly self contained and shouldn't affect anything else.

### CODE REVIEW

File: source/PrintData.cpp [hunk 1]
```diff
@@ -648,6 +665,75 @@ namespace {
...
+				break;
+			}
+
+		if(!phrase)
+		{
+			cout << "Error: invalid phrase name: \"" + phraseName + "\".\n";
+			return;
+		}
+
+		cout << "Computing results for phrase: \"" << phrase->Name() << "\"." << endl;
```
Comment (quyykk) [inline source/PrintData.cpp:699]: These messages seem to be unnecessary, and would only be annoying for anybody that wants to parse the output. So I say let's remove them.

File: source/PrintData.cpp [hunk 2]
```diff
@@ -648,6 +665,75 @@ namespace {
...
+		"-w",
+		"--weapons",
+		"-e",
+		"--engines",
+		"--power",
+		"-o",
+		"--outfits"
+	};
+
+	const set<string> VALID_ARGS()
```
Comment (quyykk) [inline source/PrintData.cpp:723]: Make this a variable instead like above, no need for this to be a function.

**Comment by bene-dictator** [comment]
Would it be possible to have a command to just generate a select number of random hails from a hail set or am I asking for too much? It would be a bit more useful for me this way, because even friendly militia 4 is a bit of a monster to look through and figure out the edge cases. It might also be a good idea for someone (it's probably going to be me) to start organising friendly civilian hails with top-level ones too :)

This is pretty useful anyway because I found some bad wording in the friendly militia 4 hail set that I should have noticed before!

**Comment by warp-core** [comment]
I can make it just acquire some number of phrases the normal way as well, yes.

**Comment by Ferociousfeind** [comment]
Any possibility to generate the number or possibilities, instead of printing all of them? So that "friendly civilian" prints 36,740 or whatever

**Comment by Ferociousfeind** [comment]
`If Parliament would listen to us and look at what pirates has done to our system, many lives would be saved.` (and similar)

found a typoooooo

gonna submit a PR to adjust thiiiiiiiis

**Comment by RisingLeaf** [comment]
> `If Parliament would listen to us and look at what pirates has done to our system, many lives would be saved.` (and similar)
> 
> found a typoooooo
> 
> gonna submit a PR to adjust thiiiiiiiis

I see we found another quyyyykkkkk

File: source/Phrase.cpp [hunk 3]
```diff
@@ -93,6 +93,89 @@ string Phrase::Get() const
...
+				choiceVector.emplace_back();
+
+				// Iterate over each element in this choice.
+				// The elements in a choice are concatenated.
+				// Each element is either empty, a string, or a pointer to another phrase.
+				for(const auto &element : choice)
+				{
+					// If this element is a string, append it to each of the partial results for this choice.
+					if(!element.first.empty())
+						AppendToVector(choiceVector, element.first);
```
Comment (RisingLeaf) [inline source/Phrase.cpp:151]: This gets used only once, is it really necessary to use a lambda here?

**Comment by a-random-lemurian** [comment]
Running `wc -l` on `"friendly militia" yields 63363 lines. 63,630 possible hails for the militia.

```
$ wc -l ~/Downloads/friendly.militia.4.txt                                                                                                                                
63363 /home/user/Downloads/friendly.militia.4.txt
```

**Comment by warp-core** [comment]
image

@bene-dictator what did you do?

**Comment by bene-dictator** [comment]
I'm only responsible for one hail for the Merchants. The rest isn't my problem :)

Also this would be very useful for ship name generation too.

**Comment by warp-core** [comment]
Ship names are all derived from phrases.

File: source/Phrase.cpp (hunk 3 above)
Comment (warp-core) [reply source/Phrase.cpp]: Done

File: source/PrintData.cpp (hunk 1 above)
Comment (warp-core) [reply source/PrintData.cpp]: Done

**Comment by warp-core** [comment]
Printing each result as it is generated in Phrase would require including `iostream` in Phrase.cpp, which seems undesirable.
I suppose we could include `PrintData.h`, which doesn't have any dependencies of its own, and add a method that can be passed strings and will print them to STDOUT. Would that be acceptable?

**Comment by quyykk** [comment]
Why not do what I wrote? :)

Add a `std::function` parameter to `Phrase::GetAll`, and then for each complete phrase you call that function. That way, for printing you can call `GetAll` with a function that prints the phrase.

**Comment by warp-core** [comment]
~~Because I didn't understand what you meant.~~
I'll try passing a function.

**Review by RisingLeaf — APPROVED** [review]
I am not really good with lambdas so I am not sure about most stuff here, but what I understand looks good. Some comments seem a bit long, might insert a linebreak there.

File: source/PrintData.cpp [hunk 4]
```diff
@@ -740,4 +839,7 @@ void PrintData::Help()
 	cerr << "    --matches: prints a list of all planets and systems matching a location filter passed in STDIN."
 			<< endl;
 	cerr << "        The first node of the location filter should be `location`." << endl;
+	cerr << "    --phrase <phrase>: prints a list of all possible results of the named phrase." << endl;
```
Comment (warp-core) [inline source/PrintData.cpp:842]:
```suggestion
	cerr << "    --phrase <phrase>: prints a list of all possible results of the named phrase." << endl;
	cerr << "        WARNING: this command may require large amounts of memory for phrases with many possible results." << endl;
	cerr << "        Use with caution. If insufficient memory is available, the computer may crash." << endl:
```

**Review by quyykk — APPROVED** [review]
LGTM (fix the merge conflicts though)

**Comment by warp-core** [comment]
Closing this until I make it better.

**Pull request closed without merging**